### ISSUE TITLE
Loss not calculated correctly when using GSPO (`importance_sampling_level == "sequence"`)

### ISSUE THREAD

**Issue by avishaiElmakies** [huggingface/trl#3823]
### Reproduction

```python
from trl import GRPOConfig

training_args = GRPOConfig(
    importance_sampling_level="sequence",
    loss_type="grpo",
    beta=0.00,  
    epsilon=3e-4,
)
```

I believe that currently the GSPO loss is calculated in a different manner than in the paper. If i am misunderstanding something, please let me know.

The loss in the paper is:

<img width="1058" height="331" alt="Image" src="https://github.com/user-attachments/assets/5ccba97d-8133-4977-a519-301786e9da15" />

which means that each example has only one Advantage.

The current code is:
https://github.com/huggingface/trl/blob/dbbc770f45139b71ecb8a2871b56db1f3ca559d0/trl/trainer/grpo_trainer.py#L1859-L1896

if we use `importance_sampling_level == "sequence"` we get that `per_token_loss` shape is ` (B, 1) `. which makes sense. but when we go to calculate the loss, we multiply `per_token_loss` with `completion_mask` which gives us a shape of ` (B, T) ` (beacuse of broadcasting) which means that the advantage is multiplied by the number of non-padding tokens for each sample. and then we sum/mean to get the loss.

e.g:

<img width="857" height="269" alt="Image" src="https://github.com/user-attachments/assets/a66f564e-8174-4bc0-9737-cdb8fe14e7bb" />

as far as i understand this is not how the calculation should be. and when we use GSPO we should just use `per_token_loss.mean() ` as the loss.

I think a similar issue arises if we use kl token loss. which will also have this broadcasting.

Maybe I am misunderstanding something, if that is the case. would love an explanation!

@qgallouedec  

### System Info

- Platform: Linux-5.14.0-503.21.1.el9_5.x86_64-x86_64-with-glibc2.34
- Python version: 3.11.12
- TRL version: 0.19.0.dev0+8f490b3
- PyTorch version: 2.7.0
- accelerator(s): NVIDIA A100-SXM4-80GB
- Transformers version: 4.53.0.dev0
- Accelerate version: 1.7.0
- Accelerate config: not found
- Datasets version: 3.6.0
- HF Hub version: 0.33.4
- bitsandbytes version: 0.46.0
- DeepSpeed version: not installed
- Diffusers version: not installed
- Liger-Kernel version: 0.5.10
- LLM-Blender version: not installed
- OpenAI version: 1.90.0
- PEFT version: 0.15.2
- vLLM version: 0.9.2

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by pramodith** [comment]
@avishaiElmakies you are correct! I'm going to open up a PR to address this issue.

**Comment by qgallouedec** [comment]
That why you need to use `loss_type="grpo"`, see the doc: https://huggingface.co/docs/trl/en/paper_index

**Comment by avishaiElmakies** [comment]
@qgallouedec, while i do understand that in the example. it should probably be noted more clearly in the docs, or maybe even in code.

**Comment by qgallouedec** [comment]
Hey @avishaiElmakies, sorry for the very late reply, catching up on the backlog.

You're right that this is a real footgun: when `importance_sampling_level="sequence"` is combined with a loss type that sums per-token contributions (`bnpo`, the new default `dapo`, `dr_grpo`, `cispo`), the broadcasted `(B, 1)` per-token loss times a `(B, T)` mask effectively length-weights each sequence's contribution, not the geometric-mean-of-per-sequence loss GSPO prescribes. The recommended combo for GSPO (`importance_sampling_level="sequence"` + `loss_type="grpo"`) is documented at https://huggingface.co/docs/trl/en/paper_index but it's silent in code, which is the gap you're pointing at.

I'd add a warning that mirrors the existing `luspo`/`vespo` mismatch checks (see https://github.com/huggingface/trl/blob/a9993736c2250da0b3d2f206ec217f144b891e5a/trl/trainer/grpo_trainer.py#L606-L616). PR welcome from anyone who wants to take it.

**Comment by discobot** [comment]
While putting the fix together, I noticed the existing
`test_train_sequence_importance_sampling` exercises exactly this footgun combo
(sequence-level importance sampling with the default `dapo`), so even the in-repo GSPO
training test was length-weighting sequence losses. Fix in #6042 — the warning mirrors the
`luspo`/`vespo` checks as suggested above and fires in that test's setup too.
